akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolMasterActor.scala

```diff
@@ -0,0 +1,140 @@
+
+  private[this] var poolStatus = Map[PoolGateway, PoolInterfaceStatus]()
+  private[this] var poolInterfaces = Map[ActorRef, PoolGateway]()
+
+  /**
+   * Start a new pool interface actor, register it in our maps, and watch its death.
+   *
+   * @param gateway the pool gateway this pool corresponds to
+   * @param fm the materializer to use for this pool
+   * @return the newly created actor ref
+   */
+  private[this] def startPoolInterfaceActor(gateway: PoolGateway)(implicit fm: Materializer): ActorRef = {
+    val props = Props(new PoolInterfaceActor(gateway)).withDeploy(Deploy.local)
+    val ref = context.system.actorOf(props, PoolInterfaceActor.name.next())
+    poolStatus += gateway -> PoolInterfaceRunning(ref)
+    poolInterfaces += ref -> gateway
+    context.watch(ref)
+  }
+
+  def receive = {
+
+    // Start or restart a pool without sending it a request. This is used to ensure that
+    // freshly created pools will be ready to serve requests immediately.
+    case s @ StartPool(gateway, materializer) ⇒
+      poolStatus.get(gateway) match {
+        case Some(PoolInterfaceRunning(_)) ⇒
+        case Some(PoolInterfaceShuttingDown(shutdownCompletedPromise)) ⇒
+          shutdownCompletedPromise.future.foreach(_ ⇒ self ! s)
+        case None ⇒
+          startPoolInterfaceActor(gateway)(materializer)
+      }
+
+    // Send a request to a pool. If needed, the pool will be started or restarted.
+    case s @ SendRequest(gateway, request, responsePromise, materializer) ⇒
+      poolStatus.get(gateway) match {
+        case Some(PoolInterfaceRunning(ref)) ⇒
+          ref ! PoolRequest(request, responsePromise)
+        case Some(PoolInterfaceShuttingDown(shutdownCompletedPromise)) ⇒
+          // The request will be resent when the pool shutdown is complete (the first
+          // request will recreate the pool).
+          shutdownCompletedPromise.future.foreach(_ ⇒ self ! s)
+        case None ⇒
+          startPoolInterfaceActor(gateway)(materializer) ! PoolRequest(request, responsePromise)
+      }
+
+    // Shutdown a pool and signal its termination.
+    case Shutdown(gateway, shutdownCompletedPromise) ⇒
+      poolStatus.get(gateway).foreach {
+        case PoolInterfaceRunning(ref) ⇒
+          ref ! PoolInterfaceActor.Shutdown
+          poolStatus += gateway -> PoolInterfaceShuttingDown(shutdownCompletedPromise)
+        case PoolInterfaceShuttingDown(formerPromise) ⇒
+          shutdownCompletedPromise.completeWith(formerPromise.future)
+        case _ ⇒
+          shutdownCompletedPromise.success(Done)
+      }
+
+    // Shutdown all known pools and signal their termination.
+    case ShutdownAll(shutdownCompletedPromise) ⇒
+      val allGateways = poolStatus.keys
+      val counter = new AtomicInteger(allGateways.size)
+      for (gateway ← allGateways)
+        gateway.shutdown().foreach(_ ⇒ if (counter.decrementAndGet() == 0) shutdownCompletedPromise.success(Done))
```

Comment:
No need to clear the maps: we just start the shutdown of the various gateways, and when it is done the maps will be cleared of the corresponding entries.

The alternate way you're proposing to shutdown the gateways waits for one pool to be shutdown before going to the next. This is not a good idea, as a pool might take a long time to shutdown if it is handling ongoing connections.